### PR TITLE
CGroup Handling Enhancements

### PR DESCRIPTION
**Pull request by mheon**
Add validation to CGroup parents passed into libpod. If we are using the systemd CGroup manager, the basename must end in `.slice`; if we are using cgroupfs, the path cannot contain `cgroup.slice`. This is based on the CRI-O validation code located at https://github.com/kubernetes-incubator/cri-o/blob/master/server/sandbox_run.go#L379-L400

Also, add our CGroup path to the OCI spec, passing it into `runc`. When the container is deleted in `runc`, it will clean up configured CGroups, solving our problem of leaving CGroup scopes around after containers exit.

Closes: #497 
Closes: #496

### CODE REVIEW

**Comment by mheon** [comment]
@baude If you get a chance, can you validate that this doesn't break our stats code? I'm pretty sure it's still working, but you've worked with it more than I have.

File: libpod/runtime_ctr.go [hunk 1]
```diff
@@ -50,6 +51,24 @@ func (r *Runtime) NewContainer(rSpec *spec.Spec, options ...CtrCreateOption) (c
...
+	case CgroupfsCgroupsManager:
+		if ctr.config.CgroupParent == "" {
+			ctr.config.CgroupParent = CgroupfsDefaultCgroupParent
+		} else if strings.HasSuffix(path.Base(ctr.config.CgroupParent), ".slice") {
+			return nil, errors.Wrapf(ErrInvalidArg, "systemd slice received as cgroup parent when using cgroupfs")
+		}
+	case SystemdCgroupsManager:
+		if ctr.config.CgroupParent == "" {
+			ctr.config.CgroupParent = SystemdDefaultCgroupParent
+		} else if len(ctr.config.CgroupParent) < 6 || !strings.HasSuffix(path.Base(ctr.config.CgroupParent), ".slice") {
```
Comment (TomSweeneyRedHat) [inline libpod/runtime_ctr.go]: I know just enough about cgroups to be dangerous.  Should ".slice" here be "system.slice"?  Could a cgroups.slice string satisfy this and I'm not sure you'd want that?  @rhatdan?

Comment (mheon) [reply libpod/runtime_ctr.go]: I'm pretty sure that anything with .slice is systemd-managed, which is why we want to avoid them when using cgroupfs - but not 100%. For what it's worth, this code is basically identical to CRI-O's validation.

Comment (rhatdan) [reply libpod/runtime_ctr.go]: Yes .slice means systemd.

**Comment by mheon** [comment]
Seeing some issues with cgroup parent, and a 'podman run' flag. Chasing them down.

**Comment by mheon** [comment]
Fixed CGroup parent. Unclear if the 'podman run' issue is actually a bug or a flake.

**Comment by mheon** [comment]
bot, retest this please

**Comment by mheon** [comment]
bot, retest this please

**Comment by mheon** [comment]
bot, retest this please

**Comment by nathwill** [comment]
is this going to break running a container under podman from a systemd unit like the one below, or would we just need to delegate differently?

```
[Unit]
Description=CRI-O container: %p

[Service]
ExecStartPre=-/bin/podman stop %p
ExecStartPre=-/bin/podman rm %p
ExecStart=/bin/podman run --net=host --workdir=/data --volume=/data:/data --log-driver=json-file --log-opt=path=/var/log/redis.log \
    --cidfile=/var/run/%p.crio --cgroup-parent=/machine.slice/%p.service \
    --name=%p redis:3.2 
ExecStop=/bin/podman stop %p
ExecStop=/bin/podman rm %p
Restart=always
Slice=machine.slice

[Install]
WantedBy=multi-user.target
```

i've been using `--cgroup-parent=/machine.slice/%p.service` to make sure all contained processes end up in the service unit's cgroup for tracking (e.g. `systemctl status`), and for ensuring full termination on container stop.

it's working pretty well so far:

```
[root@default-centos-7 ~]# systemd-cgls --no-pager /machine.slice/redis.service
/machine.slice/redis.service:
├─3812 /bin/podman run --net=host --workdir=/data --volume=/data:/data --log-driver=json-file --log-opt=path=/var/log/redis.log --cidfile=...
└─libpod-conmon-3594f3290709d0a764d307c4d4f95e62f17e83fd9c2ff2b3717209100ec44d37
  ├─3868 /usr/libexec/podman/conmon -c 3594f3290709d0a764d307c4d4f95e62f17e83fd9c2ff2b3717209100ec44d37 -u 3594f3290709d0a764d307c4d4f95e6...
  └─3594f3290709d0a764d307c4d4f95e62f17e83fd9c2ff2b3717209100ec44d37
    └─3877 redis-server *:6379
```

i'm using`machine.slice` in anticipation of hooks support landing so units like the above can be used with oci-register-machine to fully integrate the contained processes as services/machines with the host and apply resource-constraints to prioritize system stuff over "workload" services.

**Comment by mheon** [comment]
@nathwill I've been digging around under the hood, and I'm fairly certain our cgroup handling for systemd-managed cgroups is very wrong right now - we're inserting ourself into their hierarchy without telling systemd. It seems to be working, but it is definitely not the right way of doing things and is very fragile.

This patch as written would break using the systemd cgroup hierarchy in an effort to fix cgroupfs, but I think we can get both working with a bit more code. I'll mark this WIP and take a stab at it on Monday.

Be aware that you'll probably need to modify the configuration file or set a flag to use the systemd cgroup manager after these changes land, though. The config change should be a one-liner, but we should call it out in the manpage/readme.

**Comment by mheon** [comment]
An update on this: I've enabled proper CGroup handling via the runc systemd cgroup manager, based on the CRI-O code for the same. However, CRI-O places Conmon and the container's own processes in two separate CGroups. This doesn't seem like a particularly nice way of doing things (we get everything under the same CGroup with cgroupfs), so I'm going to talk to the CRI-O devs to see if there is a reason things are done that way.

**Comment by mheon** [comment]
bot, retest this please

**Comment by mheon** [comment]
bot, retest this please

**Comment by nathwill** [comment]
fwiw, i tested this out, and things seem to work as intended 👏. 

i was also able to get it working from a systemd service unit, but wanted to share a couple of observations on the experience.

cgroupfs: this PR doesn't seem to have broken the method i had been using above (i.e. using cgroupfs as the cgroup-manager and specifying the cgroup-parent as e.g. /machine.slice/redis.service).this seems to work quite well, though it sounds as though perhaps it's not intended to? (more questions on this at the end)

systemd: when specifying systemd as the cgroup-manager, i'm not able to specify the systemd-generated service cgroup (e.g. `/machine.slice/redis.service`) as the cgroup-parent, since podman only takes arguments like "xxxxx.slice". 

when doing so, podman creates libpod-*.scope as children of the specified slice, rather than the systemd-generated cgroup, as below:

```
├─machine.slice
│ ├─libpod-0411a2a82311fbc65b137ca550d5f7e6b6863bc734759be2f515b0b67062d49a.scope
│ │ └─19587 redis-server *:637
│ ├─libpod-conmon-0411a2a82311fbc65b137ca550d5f7e6b6863bc734759be2f515b0b67062d49a.scope
│ │ └─19579 /usr/libexec/podman/conmon -s -c 0411a2a82311fbc65b137ca550d5f7e6b6863bc734759be2f515b0b67062d49a -u 0411a2a82311fbc65b137ca550d5
│ └─redis.service
│   └─19564 /bin/podman --cgroup-manager=systemd run --net=host --volume=/data:/data --log-driver=json-file --log-opt=path=/var/log/redis.log
```

as a result, the conmon/container processes are un-tracked by the service unit and in order for the service to successfully start, you have to do one of:
 - use Type=oneshot/RemainAfterExit=yes in the service unit
 - skip `--detach` so that `podman run` is tracked as the MainPid
 - move `podman run` to ExecStartPre and use `podman wait` as the ExecStart.

which is more like the `docker run` experience i'm trying to get away from.

in short, `--cgroup-manager=systemd` actually ends up feeling like it integrates less with systemd than `--cgroup-manager=cgroupfs`. i know this is probably down to runc rather than podman, but thought i'd share my impressions in case they're useful.

my preference given the above would be to continue using cgroupfs. if the sole concern with cgroupfs is:

> we're inserting ourself into their hierarchy without telling systemd

would specifying [`Delegate=yes`](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#Delegate=) reasonably be expected to resolve the risk?

thanks in advance for your time and consideration! 🙏

**Comment by mheon** [comment]
@nathwill I was talking with some of the CRI-O team, and it appears we're missing some CGroup generation logic. We should be grouping the two scopes (conmon and the rest of the container) under a service we create under the parent (individual .service for containers outside a pod, or a shared one for containers in a pod). We're missing a step in the current implementation I pushed (creating that service). Would this work better for what you're doing? You'd lose the ability to name the .service you drop in `machine.slice` (we'd presumably procedurally generate it) but it would otherwise act fairly similar to what you're doing now with cgroupfs

**Comment by mheon** [comment]
Also, this still has some test failures related to kernel memory limits, specifically on CentOS. Need to look into that.

**Comment by nathwill** [comment]
> You'd lose the ability to name the .service you drop in machine.slice 

That would definitely break how I'm using it, as the service cgroup name is prescribed by systemd from the unit name, and is relied on heavily by systemd for tracking the service processes.

i totally understand the need for grouping in the pod use case, and wouldn't expect this to work for grouping service-launched containers into the same pod, but it'd be super helpful to have an option to specify the service cgroup name for non-pod containers in order to preserve the ability to use podman to run containers from service units that are able to have all the resultant container processes be tracked by systemd.

if the plan is to put all containers into pods by default, could the service cgroup could be more directly based on the pod name when provided (e.g. `podman --cgroup-manager=systemd run --pod redis --cgroup-parent=machine.slice redis:3.2` => `/machine.slice/redis.service` cgroup path)?

File: libpod/container.go [hunk 2]
```diff
@@ -39,8 +38,12 @@ const (
...
 )
 
-// DefaultCgroupParent is the default prefix to a cgroup path in libpod
-var DefaultCgroupParent = "/libpod_parent"
+// CgroupfsDefaultCgroupParent is the cgroup parent for CGroupFS in libpod
+const CgroupfsDefaultCgroupParent = "/libpod_parent"
+
+// SystemdDefaultCgroupParent is the cgroup parent for the systemd cgroup
+// manager in libpod
+const SystemdDefaultCgroupParent = "system.slice"
```
Comment (nathwill) [inline libpod/container.go]: any particular reason for using system.slice here? my understanding from the [reading](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/resource_management_guide/sec-default_cgroup_hierarchies) i've done was that machine.slice is the preferred target for containers ("machines" being interpreted here as referring to processes launched from rootfs tarballs like OCI or machinectl supported formats, instead of normal "contained" system/user processes).

my worry with a system.slice default is namespace conflicts with system services running under system.slice. as e.g. if my pod scheduler is running etcd.service on the host, and receives an etcd pod as part of an application deployment, do they both end up under /system.slice/etcd.service? since machine.slice is unpopulated by default and intended as a home for containers, it seems like a preferable default cgroup parent, unless i'm missing something?

**Comment by mheon** [comment]
Hm. We can probably wire out an option to allow setting service name (or potentially detect if there is a `.service` in the cgroup parent?). I'll look into this more.

**Comment by rhatdan** [comment]
@mrunalp PTAL

**Comment by nathwill** [comment]
fwiw, in the latest podman, using --conmon-pidfile when running podman containers as systemd services is working quite well, even without specifying --cgroup-parent. hopefully that simplifies things a bit :)

**Comment by rhatdan** [comment]
@mheon needs a rebase.

**Comment by mheon** [comment]
Alright. This should fix cgroupfs CGroup handling, which is increasingly becoming a problem. Systemd-based cgroups are still broken, but can be fixed later. I'm going to try and push this through today.

**Comment by rhatdan** [comment]
Actually this might fix another issue we are seeing with systemd, where the cgroup of the container ends up matching the users and systemd, then kills off the user session.

**Comment by mheon** [comment]
@rhatdan That's why I'm trying to get it in :smile:

**Comment by mheon** [comment]
Alright, tests appear to be passing. There are definitely still issues around systemd cgroups, but this resolves our cgroupfs issues.
@rhatdan @baude @TomSweeneyRedHat PTAL

**Comment by mheon** [comment]
Nevermind, this doesn't actually clean up cgroups. Separating the libpod and conmon CGroups fixed the tests, but the conmon cgroup hierarchy remains after the container exits.
Potentially can add some code to container.Cleanup() to handle this

**Comment by aalba6675** [comment]
@mheon  I've tested this branch with mounts: I'm seeing a stat for a duplicate path failing

```cgroup on /sys/fs/cgroup/systemd/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr```

```
unable to start container "bobby_silver": container create failed: container_linux.go:348: starting container process caused "process_linux.go:402: container init caused \"rootfs_linux.go:58: mounting \\\"cgroup\\\" to rootfs \\\"/var/lib/containers/storage/overlay/52f7959a1a8a171b2c8aee587ea81c964e84130681444f0ff03b3202804a91cb/merged\\\" at \\\"/sys/fs/cgroup\\\" caused \\\"stat /sys/fs/cgroup/systemd/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr: no such file or directory\\\"\""
```

**Comment by mheon** [comment]
@aalba6675 Damn. I was a little optimistic it would fix this, too. Are you using the same reproducer as #730

**Comment by mheon** [comment]
This does confirm that something is doubling our CGroup path, though (as opposed to pasting another type of path after it). I continue to suspect it's not us, but an OCI hook, given that the method we use to compute CGroup path never changes.

**Comment by mheon** [comment]
Cleanup code implemented. Assuming tests are green, this is finally good for review and merge.

Systemd cgroup manager code is next.

**Comment by mheon** [comment]
Alright, this is finally ready. @rhatdan PTAL

File: docs/podman.1.md [hunk 3]
```diff
@@ -23,6 +23,9 @@ has the capability to debug pods/images created by crio.
 **--help, -h**
   Print usage statement
 
+**--cgroup-manager**
```
Comment (TomSweeneyRedHat) [inline docs/podman.1.md]: A blog showing how this works would probably be nice once this hits the streets.

**Comment by aalba6675** [comment]
I am still seeing doubled cgroup scopes with a container+bind mounts.

By the second or third start/stop cycle there are 86(!) mounts of this form. There are only 3 distinct mounts but, i.e., 

mount reports 86 cgroup mounts  on
```
**/ctr/**
**/ctr/**/ctr/**
**/ctr/**/ctr/**/ctr/**
```
Of these 86 mounts only 3 are unique. 

```
cgroup on /sys/fs/cgroup/systemd/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr type cgroup (rw,nosuid,nodev,noexec,relatime,seclabel,xattr,name=systemd)
cgroup on /sys/fs/cgroup/systemd/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr type cgroup (rw,nosuid,nodev,noexec,relatime,seclabel,xattr,name=systemd)
cgroup on /sys/fs/cgroup/systemd/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr type cgroup (rw,nosuid,nodev,noexec,relatime,seclabel,xattr,name=systemd)
cgroup on /sys/fs/cgroup/systemd/libpod_parent/libpod-cd8be22a52efaed7e2790d2eb3421c00542c3eb9763bfe715c3ad23647c419e0/ctr type cgroup (rw,nosuid,nodev,noexec,relatime,seclabel,xattr,name=systemd)
```

**Comment by mheon** [comment]
@aalba6675 I strongly suspect the bug there (and the required fix) are in `oci-systemd-hook`. Our internal CGroup handling follows a strict pattern for generating supported paths and will never append them. Additionally, we don't directly mount CGroups into containers, that must be requested by something else (and the only thing I know of that does this would be `oci-systemd-hook`).

**Comment by rhatdan** [comment]
@aalba6675 could you try without using a container that runs init, and see if the issue goes away.

I believe that there is a bug in oci hooks in runc, that is not calling oci-systemd-hoop in the posthook phase.  This is why the cgroup hooks are being left behind.

**Comment by aalba6675** [comment]
@rhatdan - yes it goes away: running a /bin/bash container with fedora:28 image there is no cgroup cruft left behind.

There are also no mounts of type cgroup on /sys/fs/cgroup/systemd/libpod_parent/...

**Comment by mheon** [comment]
The lack of mounts is expected - we need those for containers running `systemd` as PID 1 (`systemd` will want to manage its own cgroups within the container). Normal containers don't need to change resource constraints, so we don't need the mounts there.

This seems to confirm it's a hook, though. The theory from @rhatdan that this is a postrun hook failing to fire and clean things up seems to make more sense than my initial bug in the hook theory, but we can't say for sure without looking into it more.

**Comment by baude** [comment]
LGTM

**Comment by baude** [comment]
@rh-atomic-bot r+
